File: lib/src/services/log_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math' as math;
import 'package:flutter/widgets.dart';
import 'package:http/http.dart' as http;
import '../models/log_entry.dart';
import '../constants/rivium_trace_constants.dart';
import 'rivium_trace_logger.dart';

/// Service for batching and sending logs to RiviumTrace
///
/// Features (matching Better Stack/Logtail):
/// - Lazy timer: only runs when buffer has logs
/// - Exponential backoff: retries with increasing delays (1s, 2s, 4s, 8s...)
/// - Max buffer size: drops oldest logs when buffer exceeds limit
/// - Lifecycle hooks: flushes on app background, pauses when inactive
class LogService with WidgetsBindingObserver {
  final String apiKey;
  final String apiUrl;
  final String? sourceId;
  final String? sourceName;
  final String platform;
  final String environment;
  final String? release;
  final int batchSize;
  final Duration flushInterval;
  final int maxBufferSize;
  final http.Client _httpClient;

  final List<LogEntry> _buffer = [];
  Timer? _flushTimer;
  bool _isFlushing = false;
  int _retryAttempt = 0;
  bool _isAppActive = true;

  // Exponential backoff constants
  static const Duration _baseRetryDelay = Duration(seconds: 1);
  static const Duration _maxRetryDelay = Duration(seconds: 60);
  static const int _maxRetryAttempts = 10;

  LogService({
    required this.apiKey,
    String? apiUrl,
    this.sourceId,
    this.sourceName,
    required this.platform,
    required this.environment,
    this.release,
    this.batchSize = 50,
    this.flushInterval = const Duration(seconds: 30),
    this.maxBufferSize = 1000, // Drop oldest logs when exceeding this
    http.Client? httpClient,
  }) : apiUrl = apiUrl ?? RiviumTraceConstants.apiUrl,
       _httpClient = httpClient ?? http.Client() {
    // Register lifecycle observer
    WidgetsBinding.instance.addObserver(this);
  }

  /// Handle app lifecycle changes
  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    switch (state) {
      case AppLifecycleState.paused:
      case AppLifecycleState.inactive:
      case AppLifecycleState.detached:
        // App going to background - flush immediately
        _isAppActive = false;
        flush();
        break;
      case AppLifecycleState.resumed:
        // App returning to foreground - resume normal operation
        _isAppActive = true;
        _retryAttempt = 0; // Reset retry count
        if (_buffer.isNotEmpty) {
          _scheduleFlush();
        }
        break;
      case AppLifecycleState.hidden:
        _isAppActive = false;
        break;
    }
  }

  /// Calculate retry delay with exponential backoff
  Duration _getRetryDelay() {
    final delay = _baseRetryDelay * math.pow(2, _retryAttempt).toInt();
    return delay > _maxRetryDelay ? _maxRetryDelay : delay;
  }

  /// Start or restart the flush timer (only if buffer has logs)
  void _scheduleFlush() {
    // Cancel existing timer
    _flushTimer?.cancel();

    // Don't schedule if app is inactive
    if (!_isAppActive) return;

    // Only schedule if there are logs to send
    if (_buffer.isNotEmpty) {
      // Use exponential backoff delay if retrying, otherwise normal interval
      final delay = _retryAttempt > 0 ? _getRetryDelay() : flushInterval;

      _flushTimer = Timer(delay, () {
        flush();
      });
    }
  }

  /// Enforce max buffer size by dropping oldest logs
  void _enforceMaxBufferSize() {
    if (_buffer.length > maxBufferSize) {
      final dropCount = _buffer.length - maxBufferSize;
      _buffer.removeRange(0, dropCount);
      RiviumTraceLogger.warning('Buffer overflow: dropped $dropCount oldest logs');
    }
  }

  /// Add a log entry to the buffer
  void add(LogEntry entry) {
    _buffer.add(entry);

    // Enforce max buffer size (drop oldest if exceeds limit)
    _enforceMaxBufferSize();

    // Auto-flush if buffer is full
    if (_buffer.length >= batchSize) {
      flush();
    } else if (_flushTimer == null || !_flushTimer!.isActive) {
      // Schedule flush only if timer isn't already running
      _scheduleFlush();
    }
  }

  /// Add a log with convenience parameters
  void log(
    String message, {
    LogLevel level = LogLevel.info,
    Map<String, dynamic>? metadata,
    String? userId,
  }) {
    add(LogEntry(
      message: message,
      level: level,
      timestamp: DateTime.now(),
      metadata: metadata,
      userId: userId,
    ));
  }

  /// Send a single log immediately (bypasses batching)
  Future<bool> sendImmediate(LogEntry entry) async {
    try {
      final payload = {
        ...entry.toJson(),
        'platform': platform,
        'environment': environment,
        if (release != null) 'release': release,
        if (sourceId != null) 'sourceId': sourceId,
        if (sourceName != null) 'sourceName': sourceName,
        'sourceType': 'sdk',
      };

      final response = await _httpClient.post(
        Uri.parse('$apiUrl/api/logs/ingest'),
        headers: {
          'Content-Type': 'application/json',
          'x-api-key': apiKey,
          'User-Agent': 'RiviumTrace-SDK/${RiviumTraceConstants.sdkVersion} ($platform)',
        },
        body: jsonEncode(payload),
      );

      if (response.statusCode >= 200 && response.statusCode < 300) {
        RiviumTraceLogger.debug('Log sent successfully');
        return true;
      } else {
        RiviumTraceLogger.warning('Failed to send log: ${response.statusCode}');
        return false;
      }
    } catch (e) {
      RiviumTraceLogger.error('Error sending log', e);
      return false;
    }
  }

  /// Flush all buffered logs to the server
  Future<bool> flush() async {
    // Cancel timer since we're flushing now
    _flushTimer?.cancel();
    _flushTimer = null;

    if (_buffer.isEmpty || _isFlushing) return true;

    _isFlushing = true;
    final logsToSend = List<LogEntry>.from(_buffer);
    _buffer.clear();

    try {
      // If no sourceId, send individual logs
      if (sourceId == null) {
        var allSucceeded = true;
        for (final entry in logsToSend) {
          if (!await sendImmediate(entry)) {
            allSucceeded = false;
          }
        }
        _retryAttempt = 0; // Reset on success
        return allSucceeded;
      }

      // Batch send
      final payload = {
        'sourceId': sourceId,
        if (sourceName != null) 'sourceName': sourceName,
        'sourceType': 'sdk',
        'logs': logsToSend.map((e) => {
          ...e.toJson(),
          'platform': platform,
          'environment': environment,
          if (release != null) 'release': release,
        }).toList(),
      };

      final response = await _httpClient.post(
        Uri.parse('$apiUrl/api/logs/ingest/batch'),
        headers: {
          'Content-Type': 'application/json',
          'x-api-key': apiKey,
          'User-Agent': 'RiviumTrace-SDK/${RiviumTraceConstants.sdkVersion} ($platform)',
        },
        body: jsonEncode(payload),
      );

      if (response.statusCode >= 200 && response.statusCode < 300) {
        RiviumTraceLogger.debug('Batch logs sent: ${logsToSend.length}');
        _retryAttempt = 0; // Reset on success
        return true;
      } else {
        RiviumTraceLogger.warning('Failed to send batch logs: ${response.statusCode}');
        // Put logs back in buffer for retry
        _buffer.insertAll(0, logsToSend);
        _enforceMaxBufferSize(); // Don't exceed max when re-adding
        // Increment retry attempt and schedule with backoff
        if (_retryAttempt < _maxRetryAttempts) {
          _retryAttempt++;
          _scheduleFlush();
        } else {
          RiviumTraceLogger.error('Max retry attempts reached, logs will be dropped');
          _retryAttempt = 0;
        }
        return false;
      }
    } catch (e) {
      RiviumTraceLogger.error('Error flushing logs', e);
      // Put logs back in buffer for retry
      _buffer.insertAll(0, logsToSend);
      _enforceMaxBufferSize(); // Don't exceed max when re-adding
      // Increment retry attempt and schedule with backoff
      if (_retryAttempt < _maxRetryAttempts) {
        _retryAttempt++;
        _scheduleFlush();
      } else {
        RiviumTraceLogger.error('Max retry attempts reached, logs will be dropped');
        _retryAttempt = 0;
      }
      return false;
    } finally {
      _isFlushing = false;
    }
  }

  /// Get the number of buffered logs
  int get bufferSize => _buffer.length;

  /// Dispose the service
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    _flushTimer?.cancel();
    flush(); // Try to send remaining logs
    _httpClient.close();
  }
}

File: android/src/main/kotlin/co/rivium/trace/sdk/services/LogService.kt
package co.rivium.trace.sdk.services

import android.app.Application
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ProcessLifecycleOwner
import co.rivium.trace.sdk.models.LogEntry
import co.rivium.trace.sdk.models.LogLevel
import co.rivium.trace.sdk.utils.RiviumTraceLogger
import org.json.JSONArray
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.TimeZone
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import kotlin.math.min
import kotlin.math.pow

/**
 * Service for batching and sending logs to RiviumTrace
 *
 * Features (matching Better Stack/Logtail):
 * - Lazy timer: only runs when buffer has logs
 * - Exponential backoff: retries with increasing delays (1s, 2s, 4s, 8s...)
 * - Max buffer size: drops oldest logs when buffer exceeds limit
 * - Lifecycle hooks: flushes on app background
 */
class LogService(
    private val apiKey: String,
    private val sourceId: String? = null,
    private val sourceName: String? = null,
    private val platform: String = "android",
    private val environment: String = "production",
    private val release: String? = null,
    private val batchSize: Int = 50,
    private val flushIntervalMs: Long = 30000,
    private val maxBufferSize: Int = 1000,
    apiUrl: String = "https://trace.rivium.co"
) : DefaultLifecycleObserver {
    private val buffer = CopyOnWriteArrayList<LogEntry>()
    private val isFlushing = AtomicBoolean(false)
    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
    private var flushTimer: Timer? = null
    private val timerLock = Any()
    private val retryAttempt = AtomicInteger(0)
    @Volatile private var isAppActive = true

    private val apiEndpoint = apiUrl.trimEnd('/')

    // Exponential backoff constants
    private val baseRetryDelayMs: Long = 1000
    private val maxRetryDelayMs: Long = 60000
    private val maxRetryAttempts = 10

    init {
        // Register for app lifecycle events
        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
    }

    // MARK: - Lifecycle Hooks

    override fun onStop(owner: LifecycleOwner) {
        // App going to background
        isAppActive = false
        flush(null)
    }

    override fun onStart(owner: LifecycleOwner) {
        // App returning to foreground
        isAppActive = true
        retryAttempt.set(0)
        if (buffer.isNotEmpty()) {
            scheduleFlush()
        }
    }

    /**
     * Calculate retry delay with exponential backoff
     */
    private fun getRetryDelay(): Long {
        val delay = baseRetryDelayMs * 2.0.pow(retryAttempt.get().toDouble()).toLong()
        return min(delay, maxRetryDelayMs)
    }

    /**
     * Enforce max buffer size by dropping oldest logs
     */
    private fun enforceMaxBufferSize() {
        while (buffer.size > maxBufferSize) {
            buffer.removeAt(0)
            RiviumTraceLogger.warn("Buffer overflow: dropped oldest log")
        }
    }

    /**
     * Schedule a one-shot flush timer (only if buffer has logs)
     */
    private fun scheduleFlush() {
        synchronized(timerLock) {
            flushTimer?.cancel()

            // Don't schedule if app is inactive
            if (!isAppActive) return

            // Only schedule if there are logs to send
            if (buffer.isEmpty()) return

            // Use exponential backoff delay if retrying, otherwise normal interval
            val delay = if (retryAttempt.get() > 0) getRetryDelay() else flushIntervalMs

            flushTimer = Timer("RiviumTraceLogFlush", true).apply {
                schedule(object : TimerTask() {
                    override fun run() {
                        flush(null)
                    }
                }, delay)
            }
        }
    }

    /**
     * Cancel the flush timer
     */
    private fun cancelFlushTimer() {
        synchronized(timerLock) {
            flushTimer?.cancel()
            flushTimer = null
        }
    }

    /**
     * Add a log entry to the buffer
     */
    fun add(entry: LogEntry) {
        buffer.add(entry)

        // Enforce max buffer size (drop oldest if exceeds limit)
        enforceMaxBufferSize()

        if (buffer.size >= batchSize) {
            flush(null)
        } else if (flushTimer == null) {
            // Schedule flush only if timer isn't already running
            scheduleFlush()
        }
    }

    /**
     * Add a log with convenience parameters
     */
    fun log(
        message: String,
        level: LogLevel = LogLevel.INFO,
        metadata: Map<String, Any?>? = null,
        userId: String? = null
    ) {
        add(LogEntry(
            message = message,
            level = level,
            metadata = metadata,
            userId = userId
        ))
    }

    /**
     * Send a single log immediately (bypasses batching)
     */
    fun sendImmediate(entry: LogEntry, callback: ((Boolean) -> Unit)? = null) {
        executor.execute {
            try {
                val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
                    timeZone = TimeZone.getTimeZone("UTC")
                }

                val payload = JSONObject().apply {
                    put("message", entry.message)
                    put("level", entry.level.value)
                    put("timestamp", dateFormat.format(entry.timestamp))
                    put("platform", platform)
                    put("environment", environment)
                    put("sourceType", "sdk")
                    release?.let { put("release", it) }
                    sourceId?.let { put("sourceId", it) }
                    sourceName?.let { put("sourceName", it) }
                    entry.userId?.let { put("userId", it) }
                    entry.metadata?.let { put("metadata", JSONObject(it)) }
                }

                val success = sendRequest("/api/logs/ingest", payload)
                callback?.invoke(success)
            } catch (e: Exception) {
                RiviumTraceLogger.error("Failed to send log: ${e.message}")
                callback?.invoke(false)
            }
        }
    }

    /**
     * Flush all buffered logs to the server
     */
    fun flush(callback: ((Boolean) -> Unit)?) {
        // Cancel timer since we're flushing now
        cancelFlushTimer()

        if (buffer.isEmpty() || !isFlushing.compareAndSet(false, true)) {
            callback?.invoke(true)
            return
        }

        val logsToSend = ArrayList(buffer)
        buffer.clear()

        executor.execute {
            try {
                // If no sourceId, send individual logs
                if (sourceId == null) {
                    var allSucceeded = true
                    for (entry in logsToSend) {
                        val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
                            timeZone = TimeZone.getTimeZone("UTC")
                        }

                        val payload = JSONObject().apply {
                            put("message", entry.message)
                            put("level", entry.level.value)
                            put("timestamp", dateFormat.format(entry.timestamp))
                            put("platform", platform)
                            put("environment", environment)
                            put("sourceType", "sdk")
                            release?.let { put("release", it) }
                            entry.userId?.let { put("userId", it) }
                            entry.metadata?.let { put("metadata", JSONObject(it)) }
                        }

                        if (!sendRequest("/api/logs/ingest", payload)) {
                            allSucceeded = false
                        }
                    }
                    isFlushing.set(false)
                    callback?.invoke(allSucceeded)
                    return@execute
                }

                // Batch send
                val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
                    timeZone = TimeZone.getTimeZone("UTC")
                }

                val logsArray = JSONArray()
                for (entry in logsToSend) {
                    val logObj = JSONObject().apply {
                        put("message", entry.message)
                        put("level", entry.level.value)
                        put("timestamp", dateFormat.format(entry.timestamp))
                        put("platform", platform)
                        put("environment", environment)
                        release?.let { put("release", it) }
                        entry.userId?.let { put("userId", it) }
                        entry.metadata?.let { put("metadata", JSONObject(it)) }
                    }
                    logsArray.put(logObj)
                }

                val payload = JSONObject().apply {
                    put("sourceId", sourceId)
                    put("sourceType", "sdk")
                    put("logs", logsArray)
                    sourceName?.let { put("sourceName", it) }
                }

                val success = sendRequest("/api/logs/ingest/batch", payload)

                if (success) {
                    retryAttempt.set(0) // Reset on success
                } else {
                    // Put logs back in buffer for retry
                    buffer.addAll(0, logsToSend)
                    enforceMaxBufferSize() // Don't exceed max when re-adding
                    // Increment retry attempt and schedule with backoff
                    if (retryAttempt.get() < maxRetryAttempts) {
                        retryAttempt.incrementAndGet()
                        scheduleFlush()
                    } else {
                        RiviumTraceLogger.error("Max retry attempts reached, logs will be dropped")
                        retryAttempt.set(0)
                    }
                }

                isFlushing.set(false)
                callback?.invoke(success)
            } catch (e: Exception) {
                RiviumTraceLogger.error("Failed to flush logs: ${e.message}")
                // Put logs back in buffer for retry
                buffer.addAll(0, logsToSend)
                enforceMaxBufferSize() // Don't exceed max when re-adding
                // Increment retry attempt and schedule with backoff
                if (retryAttempt.get() < maxRetryAttempts) {
                    retryAttempt.incrementAndGet()
                    scheduleFlush()
                } else {
                    RiviumTraceLogger.error("Max retry attempts reached, logs will be dropped")
                    retryAttempt.set(0)
                }
                isFlushing.set(false)
                callback?.invoke(false)
            }
        }
    }

    /**
     * Get the number of buffered logs
     */
    val bufferSize: Int
        get() = buffer.size

    /**
     * Shutdown the service
     */
    fun shutdown() {
        ProcessLifecycleOwner.get().lifecycle.removeObserver(this)
        cancelFlushTimer()
        flush(null)
        executor.shutdown()
    }

    private fun sendRequest(endpoint: String, payload: JSONObject): Boolean {
        var connection: HttpURLConnection? = null
        return try {
            val url = URL(apiEndpoint + endpoint)
            connection = url.openConnection() as HttpURLConnection
            connection.apply {
                requestMethod = "POST"
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("x-api-key", apiKey)
                connectTimeout = 30000
                readTimeout = 30000
                doOutput = true
            }

            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(payload.toString())
                writer.flush()
            }

            val responseCode = connection.responseCode
            val success = responseCode in 200..299

            if (success) {
                RiviumTraceLogger.debug("Logs sent successfully")
            } else {
                RiviumTraceLogger.warn("Failed to send logs: HTTP $responseCode")
            }

            success
        } catch (e: Exception) {
            RiviumTraceLogger.error("Error sending logs: ${e.message}")
            false
        } finally {
            connection?.disconnect()
        }
    }
}
